dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/security/config/WebMvcConfig.java

```diff
@@ -223,6 +225,7 @@ protected void addFormatters( FormatterRegistry registry )
     {
         registry.addConverter( new StringToOrderCriteriaListConverter() );
         registry.addConverter( new FieldPathConverter() );
+        registry.addConverter( new StringToSetConverter() );
```

Comment:
I am pretty sure this would cause issues. For example the `fields` parameter expects a comma separated list  of inputs to result in a list/set of the fields listed. With the replacement now I guess it would result in a set with just a single entry of all the values. We also have other parameters like the `filter` parameter where we do custom split because split by `,` would result in the wrong list. In such cases the parameter is simply modelled as `String` in the controller method and the splitting is done afterwards in dedicated code. I don't see why such a solution would not also work well in your case. If you want a `Set<String>` with just one value of "everything" just model it as `String` in the API and construct the `Set.of` from that API parameter.

Reply:
Just had a quick look at this, and I am not sure I quite follow/understand:

The core of the problem here, is that "attribute=<uid>:<op>:41,0" is split incorrectly, as we want "41,0" to the be search string, not 41 and 0 individually. It makes sense to me that we need to treat the 41,0 as a string, not a set.

As Jan mentions, we have several parameters that expects "," to separate items in a list of values, rather than being part of a single value (IE. fields, filter).

My understanding is that @jbee concern is that this change will alter the behaviour for those other parameters, since we are now changing how everything is converted, instead of just the /query?attribute param.

But now you are saying @lucaCambi77 that this change is not changing how it works? Can we remove that change and the problem is still fixed? If not, is there an alternative way to solve this that does not affect all controllers, but just the endpoint in question? (/Query)?

Reply:
`attribute=::41,0` like any parameter request is a sequence of chars. Then Spring is able to convert this to an object by using default converters.
When can have more than one `?filter` o `attributes` in a request, so we want this to be a set object. Unfortunately, Spring has a default way to do this conversion by splitting at `,` a parameter. Hence attribute=::41,0 will become a set object 

- attribute=::41
- 0

This pr fixes this by using a converter so a string won't break. This is for every controller because I don't see a reason why we need this behavior to be in place.  This is a minimal change that is not affecting the way `?filter` or ?`attributes` work.
Notice that I don't think `?fields`  will use this because it is a List and not a Set. However, `?fields` has already a converter.
If you are worrying about what this breaks, please give me an example so I understand

Reply:
Yes, @stian-sandvold is mostly right about my concerns. I missed that what we use elsewhere is `List<String>` and `Set<String>` is otherwise unused as you say @lucaCambi77 . So changing the behaviour for target type `Set<String>` would not effect existing parameters I still think changing its default to not split on `,` is not a good choice as we would expect exactly that behaviour should the target be a `List<String>` or even a `Collection<String>`. I think your use case of not wanting to split on `,` because it is part of a single value is the odd case that should not adjust the common case. Esepcially not if that than behaves inconsistent with similar cases thus be confusing for developers to have a reasonable expectation. Also since the "workaround" for your special case is very simple. Have your value be modelled as `String` parameter and wrap that into a `Set<String>` in the controller code.

Reply:
@jbee 
`Have your value be modelled as String parameter and wrap that into a Set<String> in the controller code.`
With what logic should I then create a Set in the controller? Using a split char? And what if that split char is part of the filter value?

`I think your use case of not wanting to split on, because it is part of a single value is the odd case that should not adjust the common case. `
The odd case is that we rely on Spring to split on `,`. This doesn't make sense and shouldn't be the rule. Applying changes to single controllers as @stian-sandvold is saying will create an anti-pattern because we are moving the problem somewhere else. In my opinion, it should be the opposite, so given no rule on splitting, then every controller should have a specific rule if required. Looking at tests there aren't cases where we rely on Spring default split from String to Set, but in case we can just go there and add for that specific controller. 

The change is documented in the MVC configuration so if we want to apply this to List we can do it, there aren't infinite types of collections and the use case is limited.

Notice there is also another option which is converting from String to a custom type like we are doing for `?fields`. In that case, we could convert from String to a more complex object like `QueryItem` but there are multiple cases and also there is a logic that shouldn't be in a converter

Reply:
if you have like in your example say `checks=a,b,c` and `a=thing1,thing2`, then you can't reliably split on `,` to get distinct values. Yeah, now I see also in the [documentation](https://docs.dhis2.org/en/develop/using-the-api/dhis-core-version-239/new-tracker.html#common-request-parameters). Let's see in the backend meeting if we can find a solution